ARROW-12017: [R] [Documentation] Make proper developing arrow docs (#9898)
apache/arrow: r/vignettes/developing.Rmd

Even for developers, most contributions can be accomplished with the instructions in [R-only development](#r-only-development). But if you're working on both the C++ library and the R package, the [Developer environment setup](#-developer-environment-setup) section will guide you through setting up a developer environment.

## R-only development 

Windows and macOS users who wish to contribute to the R package and
don’t need to alter the Arrow C++ library may be able to obtain a
recent version of the library without building from source. On macOS,
you may install the C++ library using [Homebrew](https://brew.sh/):

``` shell
# For the released version:
brew install apache-arrow
# Or for a development version, you can try:
brew install apache-arrow --HEAD
```

> On Windows, you can download a .zip file with the arrow dependencies from the

Reviewer: This is also true for Linux now

> [nightly repository](https://arrow-r-nightly.s3.amazonaws.com/libarrow/bin/windows/),

Reviewer: ~~Unfortunately~~ As it turns out, this URL does not return useful information (the bintray version did). Surely there's an S3 permission that can be set that would make that list (XML) with contents, but a better way is probably to encourage use of the package itself, like: 

```
> nightly <- s3_bucket("arrow-r-nightly")
> nightly$ls("libarrow/bin")
[1] "libarrow/bin/centos-7"     "libarrow/bin/centos-8"    
[3] "libarrow/bin/debian-10"    "libarrow/bin/debian-9"    
[5] "libarrow/bin/ubuntu-16.04" "libarrow/bin/ubuntu-18.04"
[7] "libarrow/bin/windows"
```

We could even add a function to download the latest binary (appropriate for your OS) and unzip it/put it in the right place. Out of scope here, just saying.

Author: > We could even add a function to download the latest binary (appropriate for your OS) and unzip it/put it in the right place. Out of scope here, just saying.

I'll add this to the todo Jira, it's a good idea and something I had struggled with a little bit myself when trying to diagnose how the nightlies are downloaded and if we can match those / do that process manually/with a development version.

brew install cmake openssl
```

#### Ubuntu
```{bash, save=run & ubuntu}
sudo apt install -y cmake libcurl4-openssl-dev libssl-dev
```

### Configure the Arrow build {.tabset}

You can choose to build and then install the Arrow library into a user-defined directory or into a system-level directory. You only need to do one of these two options. It’s recommended to make a `build` directory inside of the `cpp` directory of the Arrow git repository (it is git-ignored). 

#### Install to another directory

It is recommended that you install the arrow library to a user-level directory to be used in development. In this example we will install it to a directory called `dist` that has the same parent as our `arrow` checkout. The directory name `dist` and the location is merely a convention. It could be named or located anywhere you would like. However, note that your installation of the Arrow R package will point to this directory and need it to remain intact for the package to continue to work. This is one reason we recommend *not* placing it inside of the arrow git checkout. 

```{bash, save=run & !sys_install}
export ARROW_HOME=$(pwd)/dist
mkdir $ARROW_HOME
```

> _Special instructions on Linux:_ You will need to set `LD_LIBRARY_PATH` to the same directory as `LIB_DIR` before launching R and using Arrow. One way to do this is to add it to your profile (we use `~/.bash_profile` here, but you might need to put this in a different file depending on your setup). On macOS we do not need to do this because the macOS shared library paths are hardcoded to their locations during build time.

Reviewer: I don't think `LIB_DIR` is mentioned in this document so that's not helpful (anymore)

  -DARROW_DATASET=ON \
  -DARROW_FILESYSTEM=ON \
  -DARROW_JEMALLOC=ON \
  -DARROW_JSON=ON \
  -DARROW_PARQUET=ON \
  -DARROW_WITH_SNAPPY=ON \
  -DARROW_WITH_ZLIB=ON \
  -DARROW_INSTALL_NAME_RPATH=OFF \
  ..
```

### More Arrow features

To enable optional features including: S3 support, an alternative memory allocator, and additional compression libraries, add some or all of these flags:

``` shell
  -DARROW_MIMALLOC=ON \
  -DARROW_WITH_BROTLI=ON \
  -DARROW_WITH_BZ2=ON \
  -DARROW_WITH_LZ4=ON \
  -DARROW_WITH_SNAPPY=ON \
  -DARROW_WITH_ZLIB=ON \
  -DARROW_WITH_ZSTD=ON \
```

Other flags that may be useful:

* `-DARROW_EXTRA_ERROR_CONTEXT=ON` makes errors coming from the C++ library point to files and line numbers
* `-DBoost_SOURCE=BUNDLED` and `-DThrift_SOURCE=bundled`, for example, or any other dependency `*_SOURCE`, if you have a system version of a C++ dependency that doesn't work correctly with Arrow. This tells the build to compile its own version of the dependency from source.
* `-DCMAKE_BUILD_TYPE=debug` and `-DCMAKE_BUILD_TYPE=relwithdebinfo` can be useful for debugging, though they are both slower to compile than the default `release`.

### Build Arrow

> You can `-j#` here too to speed up compilation by running in parallel (where `#` is the number of cores you have available).

Reviewer 2: Tiny point, but took me a couple of reads to realise that this sentence means you can add `-j#` to the command below.

If you need to alter both the Arrow C++ library and the R package code, or if you can’t get a binary version of the latest C++ library elsewhere, you’ll need to build it from source too. This section discusses how to set up a C++ build configured to work with the R package. For more general resources, see the [Arrow C++ developer
guide](https://arrow.apache.org/docs/developers/cpp/building.html).

### Install dependencies {.tabset}

The Arrow C++ library will by default use system dependencies if suitable versions are found; if they are not present, it will build them during its own build process. The only dependencies that one needs to install outside of the build process are `cmake` (for configuring the build) and `openssl` if you are building with S3 support.

For a faster build, you may choose to install on the system more C++ library dependencies (such as `lz4`, `zstd`, etc.) so that they don't need to be built from source in the Arrow build. This is optional.

#### macOS
```{bash, save=run & macos}
brew install cmake openssl
```

#### Ubuntu
```{bash, save=run & ubuntu}
sudo apt install -y cmake libcurl4-openssl-dev libssl-dev
```

### Configure the Arrow build {.tabset}

> You can choose to build and then install the Arrow library into a user-defined directory or into a system-level directory. You only need to do one of these two options.

Reviewer: Let's make clear that there are effectively three steps, which involve two directories:

* configure (`cmake`), which populates a build directory with scripts etc.
* build, which compiles arrow (including potentially downloading and building dependencies) and produces the resulting libraries and headers in that build directory
* install, which organizes just the libraries and headers into a clean location, without any of the build detritus.

It's this last step, where to install to, that is the key difference in the two approaches described here. And it requires a change or two in the configuring step.

Either way, you will need to create a directory into which the C++ build will put its contents. It is recommended to make a `build` directory inside of the `cpp` directory of the Arrow git repository (it is git-ignored, so you won't accidentally check it in).

> Starting from the directory that contains your git checkout of `apache/arrow`,

Reviewer: IMO we should make all instructions relative to the top-level of the repo, so drop the leading `arrow/` below

Author: I've tried to do that, generally with the caveat that the `dist` directory that is the `ARROW_HOME` should *not* be under this directory. 

IIRC, originally I had the creation of this directory right before the config (when the tutorial would be inside the repo) but you moved it up higher in your push. If we want the `mkdir call` to not include `arrow/` we should move it to a later code chunk after we've `pushd arrow`. 

I think it's probably better lower down anyway, so unless you object I think I'll move it back down there (along with most of the paragraph above).

```{bash, save=run}
mkdir -p arrow/cpp/build
```

#### Install to another directory

> It is recommended that you install the arrow library to a user-level directory to be used in development. In this example we will install it to a directory called `dist` that has the same parent as our `arrow` checkout, but it could be named or located anywhere you would like. However, note that your installation of the Arrow R package will point to this directory and need it to remain intact for the package to continue to work. This is one reason we recommend *not* placing it inside of the arrow git checkout.

Reviewer: > It is recommended

Why is it recommended?

```{bash, save=run & !sys_install}
export ARROW_HOME=$(pwd)/dist
mkdir $ARROW_HOME
```

> _Special instructions on Linux:_ You will need to set `LD_LIBRARY_PATH` to the `lib` directory that will is under where we set `$ARROW_HOME`, before launching R and using Arrow. One way to do this is to add it to your profile (we use `~/.bash_profile` here, but you might need to put this in a different file depending on your setup). On macOS we do not need to do this because the macOS shared library paths are hardcoded to their locations during build time.

Reviewer: "if you use a shell other than `bash`"

```{bash, save=run & sys_install}
pushd arrow
pushd cpp/build
```

You’ll first call `cmake` to configure the build and then `make install`. For the R package, you’ll need to enable several features in the C++ library using `-D` flags:

```{bash, save=run & sys_install}
cmake \
  -DARROW_COMPUTE=ON \
  -DARROW_CSV=ON \
  -DARROW_DATASET=ON \
  -DARROW_FILESYSTEM=ON \
  -DARROW_JEMALLOC=ON \
  -DARROW_JSON=ON \
  -DARROW_PARQUET=ON \
  -DARROW_WITH_SNAPPY=ON \
  -DARROW_WITH_ZLIB=ON \
  -DARROW_INSTALL_NAME_RPATH=OFF \
  ..
```

`..` refers to the C++ source directory: we're in `cpp/build`, and the source is in `cpp`.

### More Arrow features

To enable optional features including: S3 support, an alternative memory allocator, and additional compression libraries, add some or all of these flags (the trailing `\` makes them easier to paste into a bash shell on a new line):

``` shell
  -DARROW_MIMALLOC=ON \
  -DARROW_WITH_BROTLI=ON \
  -DARROW_WITH_BZ2=ON \
  -DARROW_WITH_LZ4=ON \
  -DARROW_WITH_SNAPPY=ON \
  -DARROW_WITH_ZSTD=ON \
```

Other flags that may be useful:

> * `-DARROW_EXTRA_ERROR_CONTEXT=ON` makes errors coming from the C++ library point to files and line numbers

Reviewer: Since this is a developer guide, should we always recommend this on?

Author: I've not actually used it, I can add it to the list if you think it's helpful

Reviewer: Yeah I think it's helpful

* `-DBoost_SOURCE=BUNDLED` and `-DThrift_SOURCE=bundled`, for example, or any other dependency `*_SOURCE`, if you have a system version of a C++ dependency that doesn't work correctly with Arrow. This tells the build to compile its own version of the dependency from source.

> * `-DCMAKE_BUILD_TYPE=debug` and `-DCMAKE_BUILD_TYPE=relwithdebinfo` can be useful for debugging, though they are both slower to compile than the default `release`.

Reviewer: ```suggestion
* `-DCMAKE_BUILD_TYPE=debug` or `-DCMAKE_BUILD_TYPE=relwithdebinfo` can be useful for debugging. You probably don't want to do this generally because a debug build is much slower at runtime than the default `release` build.
```

_Note_ `cmake` is particularly sensitive to whitespacing, if you see errors, check that you don't have any errant whitespace around

### Build Arrow

You can add `-j#` between `make` and `install` here too to speed up compilation by running in parallel (where `#` is the number of cores you have available).

```{bash, save=run & !(sys_install & ubuntu)}

> make install

Reviewer: ```suggestion
make -j8 install
```

Just to illustrate

Author: I think the most that we could do on CI is `-j2` (or `-j4` if there is multithreading, though I'm not certain there is son the runners)

Reviewer: Ok. I wasn't thinking at the time about running this on CI. That said, I just set MAKEVARS=-j100 locally and it didn't break 🤷

Author: Oh apparently saturation is less of a concern, so maybe we can do 8 and the CI will deal with swapping between them.

```

If you are installing on linux, and you are installing to the system, you may
need to use `sudo`:

```{bash, save=run & sys_install & ubuntu}
sudo make install
```

> Note that after any change to the C++ library, you must reinstall it and

Reviewer: Maybe put this under Troubleshooting?

```r
Sys.setenv(LIBARROW_MINIMAL="false")
remotes::install_github("apache/arrow/r@bugfix", build = FALSE)
```

Developers may wish to use this method of installing a specific commit
separate from another Arrow development environment or system installation
(e.g. we use this in [arrowbench](https://github.com/ursacomputing/arrowbench) to install development versions of arrow isolated from the system install). If you already have Arrow C++ libraries installed system-wide, you may need to set some additional variables in order to isolate this build from your system libraries:

* Setting the environment variable `FORCE_BUNDLED_BUILD` to `true` will skip the `pkg-config` search for Arrow libraries and attempt to build from the same source at the repository+ref given.
* You may also need to set the Makevars `CPPFLAGS` and `LDFLAGS` to `""` in order to prevent the installation process from attempting to link to already installed system versions of Arrow. One way to do this temporarily is wrapping your `remotes::install_github()` call like so: `withr::with_makevars(list(CPPFLAGS = "", LDFLAGS = ""), remotes::install_github(...))`.

> ## How does our automated building actually work?

Reviewer: I don't think "automated building" is the right term here. Something about how all of the pieces come together? Or "what happens when you R CMD INSTALL?"

There are a number of scripts that are triggered when `R CMD INSTALL .`. For Arrow users, these should all just work without configuration and pull in the most complete pieces (e.g. official binaries that we host) so the installation process is easy. However knowing about these scripts can help troubleshoot if things go wrong in them or things go wrong in an install:

> * `configure` and `configure.win` These scripts are triggered during `R CMD INSTALL .` on non-windows and windows platforms respectively. They handle finding the Arrow library, setting up the build variables necessary, and writing the package Makevars file that is used to compile the C++ code in the R package.

Reviewer: Also note that this is a standard thing for R packages with compiled code, this is not an Arrow convention

```suggestion
* `configure` and `configure.win` These scripts are triggered during `R CMD INSTALL .` on non-Windows and Windows platforms, respectively. They handle finding the Arrow library, setting up the build variables necessary, and writing the package Makevars file that is used to compile the C++ code in the R package.
```

> * `tools/nixlibs.R` This script is called by `configure` on Linux (or on any non-windows OS with the environment variable `FORCE_BUNDLED_BUILD=true`). This sets up the build process for our bundled builds (which is the default on linux). The operative logic is at the end of the script, but it will do the following (and it will stop with the first one that succeeds and some of the steps are only checked if they are enabled via an environment variable):

Reviewer: It's not always called by configure

Author: 👍 I've clarified this. Funny story: for me this phrasing doesn't imply "always" so my first reading of your comment was that some other thing calls `nixlibs.R` and I went looking for what else calls it until coming back and realizing what you meant. Oh langauge!

  * Check if there is an already built libarrow in `arrow/r/libarrow-{version}`, use that to link against if it exists.

>   * Check if a binary is available from our hosted official builds.

Reviewer: Not official

Author: That was a bad typo!

  * Download the Arrow source and build the Arrow Library from source.
  * `*** Proceed without C++` dependencies (this is an error and the package will not work, but if you see this message you know the previous steps have not succeeded/were not enabled)

> * `inst/build_arrow_static.sh` this script builds Arrow for a bundled, static build. It is called by `tools/nixlibs.R` when the Arrow library is being built.

Reviewer: Note that this should look familiar if you've read this far in this vignette--it's cmake and install.

(One difference, for another day, is static vs shared library)

> ## Editing C++ code

Reviewer: ```suggestion
## Editing C++ code in the R package
```

> The `arrow` package uses some customized tools on top of `cpp11` to

Reviewer: Why? (Something about how we have features that are conditionally built at compile time)

Author: TIL. Are there any other reasons I should add here too?

Reviewer: I think that's it.

prepare its C++ code in `src/`. If you change C++ code in the R package,

> you will need to set the `ARROW_R_DEV` environment variable to `true`

Reviewer: Or use the makefile

(optionally, add it to your `~/.Renviron` file to persist across
sessions) so that the `data-raw/codegen.R` file is used for code
generation.

> We use Google C++ style in our C++ code. Check for style errors with

Reviewer: Honestly I'd suggest the editor plugins before asking people to run a shell script with python etc. etc.

## Running tests

Some tests are conditionally enabled based on the availability of certain
features in the package build (S3 support, compression libraries, etc.).
Others are generally skipped by default but can be enabled with environment
variables or other settings:

* All tests are skipped on Linux if the package builds without the C++ libarrow.
  To make the build fail if libarrow is not available (as in, to test that
  the C++ build was successful), set `TEST_R_WITH_ARROW=true`
* Some tests are disabled unless `ARROW_R_DEV=true`
* Tests that require allocating >2GB of memory to test Large types are disabled
  unless `ARROW_LARGE_MEMORY_TESTS=true`
* Integration tests against a real S3 bucket are disabled unless credentials
  are set in `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`; these are available
  on request
* S3 tests using [MinIO](https://min.io/) locally are enabled if the
  `minio server` process is found running. If you're running MinIO with custom
  settings, you can set `MINIO_ACCESS_KEY`, `MINIO_SECRET_KEY`, and
  `MINIO_PORT` to override the defaults.

## Github workflows

> Crossbow is our extended test suite runner [more documentation is available](https://arrow.apache.org/docs/developers/crossbow.html) but important GitHub comment commands include:

Reviewer: Restructure this more like "On a pull request, there are some actions you can trigger by commenting on the PR. We have additional CI checks that run nightly and can be requested on demand using an internal tool called [crossbow](link)."

> * `@github-actions crossbow submit -g r` for extended R CI tests

Reviewer: ```suggestion
* `@github-actions crossbow submit -g r` for all extended R CI tests
```

> * `@github-actions crossbow submit {task-name}` for running a specific task

Reviewer: ```suggestion
* `@github-actions crossbow submit {task-name}` for running a specific task. See the `r:` group definition near the beginning of the [crossbow configuration](https://github.com/apache/arrow/blob/master/dev/tasks/tasks.yml) for a list of glob expression patterns that match names of items in the `tasks:` list below it.
```

* `@github-actions autotune` will run and fix lint c++ linting errors + run R documentation (among other cleanup tasks) and commit them to the branch

> 

Reviewer: (at some point) we should also talk about docker and docker-compose

Author: I'm going to add this to a follow on (especially cause I think ultimately that probably should go on the project docs with a link here to that)

This document is a work in progress and will grow + change as the Apache Arrow project grows and changes. We have tried to make these steps as robust as possible (in fact, we even test exactly these instructions on our nightly CI to ensure they don't become stale!), but certain custom configurations might conflict with these instructions and there are differences of opinion across developers about if and what the one true way to set up development environments like this is.  We also solicit any feedback you have about things that are confusing or additions you would like to see here. Please [report an issue](https://issues.apache.org/jira/projects/ARROW/issues) if there you see anything that is confusing, odd, or just plain wrong.

## R-only development

Windows and macOS users who wish to contribute to the R package and
don’t need to alter the Arrow C++ library may be able to obtain a
recent version of the library without building from source. On macOS,
you may install the C++ library using [Homebrew](https://brew.sh/):

``` shell
# For the released version:
brew install apache-arrow
# Or for a development version, you can try:
brew install apache-arrow --HEAD
```

> On Windows and Linux, you can download a .zip file with the arrow dependencies from the
> nightly repository,
> and then set the `RWINLIB_LOCAL` environment variable to point to that
> zip file before installing the `arrow` R package. Version numbers in that

Reviewer: ```suggestion
On Windows and Linux, you can download a .zip file with the arrow dependencies from the
nightly repository.
Windows users then can set the `RWINLIB_LOCAL` environment variable to point to that
zip file before installing the `arrow` R package. On Linux, you'll need to create a `libarrow` directory inside the R package directory and unzip that file into it. Version numbers in that
```

Reviewer: (This is lame and we could do a lot better, but at least we should be factually accurate.)
